tests/returntypes/test_spsa_gradient_shot_vec_new.py

```diff
@@ -0,0 +1,1394 @@
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.probs(wires=[0, 1])
+
+        # by default all parameters are assumed to be trainable
+        with pytest.raises(
+            ValueError, match=r"Cannot differentiate with respect to parameter\(s\) {0}"
+        ):
+            spsa_grad(tape, _expand=False)
+
+        # setting trainable parameters avoids this
+        tape.trainable_params = {1, 2}
+        dev = qml.device("default.qubit", wires=2, shots=default_shot_vector)
+        tapes, fn = spsa_grad(tape, h=h_val, shots=default_shot_vector)
+
+        all_res = fn(dev.batch_execute(tapes))
+
+        assert isinstance(all_res, tuple)
+        assert len(all_res) == len(default_shot_vector)
+
+        for res in all_res:
+            assert isinstance(res, tuple)
+
+            assert isinstance(res[0], numpy.ndarray)
+            assert res[0].shape == (4,)
+
+            assert isinstance(res[1], numpy.ndarray)
+            assert res[1].shape == (4,)
+
+    @pytest.mark.parametrize("num_directions", [1, 6])
+    def test_independent_parameter(self, num_directions, mocker):
+        """Test that an independent parameter is skipped
+        during the Jacobian computation."""
+        spy = mocker.spy(qml.gradients.spsa_gradient, "generate_multishifted_tapes")
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.543, wires=[0])
+            qml.RY(-0.654, wires=[1])
+            qml.expval(qml.PauliZ(0))
+
+        dev = qml.device("default.qubit", wires=2, shots=default_shot_vector)
+        tapes, fn = spsa_grad(
+            tape, h=h_val, shots=default_shot_vector, num_directions=num_directions
+        )
+        all_res = fn(dev.batch_execute(tapes))
+
+        assert isinstance(all_res, tuple)
+        assert len(all_res) == len(default_shot_vector)
+
+        for res in all_res:
+            assert isinstance(res, tuple)
+            assert len(res) == 2
+
+            assert isinstance(res[0], numpy.ndarray)
+            assert isinstance(res[1], numpy.ndarray)
+
+        # 2 tapes per direction because the default strategy for SPSA is "center"
+        assert len(spy.call_args_list) == num_directions
+
+        # Never shift the independent parameter
+        for i in range(num_directions):
+            assert spy.call_args_list[i][0][0:2] == (tape, [0])
+            assert spy.call_args_list[i][0][2].shape == (2, 1)
+
+    def test_no_trainable_params_tape(self):
+        """Test that the correct ouput and warning is generated in the absence of any trainable
+        parameters"""
+        dev = qml.device("default.qubit", wires=2, shots=default_shot_vector)
+
+        weights = [0.1, 0.2]
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=0)
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        # TODO: remove once #2155 is resolved
```

Comment:
Probably not relevant any more? (I hope)

Reply:
#2155 is not resolved... :sweat_smile: